internal/transport/http2_server.go

```diff
@@ -401,6 +402,20 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 		s.cancel()
 		return true
 	}
+	if state.data.httpMethod != http.MethodPost {
+		t.mu.Unlock()
+		if logger.V(logLevel) {
+			logger.Warningf("transport: http2Server.operateHeaders parsed a :method field: %v which should be POST", state.data.httpMethod)
+		}
+		t.controlBuf.put(&cleanupStream{
+			streamID: streamID,
+			rst:      true,
+			rstCode:  http2.ErrCodeProtocol,
+			onWrite:  func() {},
+		})
+		s.cancel()
+		return false
+	}
 	t.maxStreamID = streamID
```

Comment:
This should go above the method check so we ensure stream IDs are increasing, inclusive of this stream.

Reply:
Ah, just read through the spec. I falsely assumed receiving EOS would "tear down" stream, making stream id reusable. Just read this in HTTP/2 spec "Stream identifiers cannot be reused".